File: pmd-java/src/test/java/net/sourceforge/pmd/lang/java/ast/ASTSwitchLabelTest.java
/**
 * BSD-style license; for more info see http://pmd.sourceforge.net/license.html
 */

package net.sourceforge.pmd.lang.java.ast;

import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.util.List;

import org.checkerframework.checker.nullness.qual.NonNull;
import org.junit.jupiter.api.Test;

import net.sourceforge.pmd.lang.java.BaseParserTest;
import net.sourceforge.pmd.lang.java.JavaParsingHelper;

class ASTSwitchLabelTest extends BaseParserTest {

    private final JavaParsingHelper java = JavaParsingHelper.DEFAULT;

    @Test
    void testDefaultOff() {
        List<ASTSwitchLabel> ops = java.getNodes(ASTSwitchLabel.class, "public class Foo {\n void bar() {\n  switch (x) {\n   case 1: y = 2;\n  }\n }\n}");
        assertFalse(ops.get(0).isDefault());
    }

    @Test
    void testDefaultSet() {
        @NonNull ASTSwitchStatement switchStmt = java.parse(SWITCH_WITH_DEFAULT).descendants(ASTSwitchStatement.class).firstOrThrow();
        assertTrue(switchStmt.hasDefaultCase());
        assertFalse(switchStmt.isExhaustiveEnumSwitch());
        assertTrue(switchStmt.getBranches().firstOrThrow().getLabel().isDefault());
        assertFalse(switchStmt.getBranches().get(1).getLabel().isDefault());
    }

    @Test
    void testExhaustiveEnum() {
        @NonNull ASTSwitchStatement switchStmt = java.parse(EXHAUSTIVE_ENUM).descendants(ASTSwitchStatement.class).firstOrThrow();
        assertFalse(switchStmt.hasDefaultCase());
        assertTrue(switchStmt.isExhaustiveEnumSwitch());
    }

    @Test
    void testNotExhaustiveEnum() {
        @NonNull ASTSwitchStatement switchStmt = java.parse(NOT_EXHAUSTIVE_ENUM).descendants(ASTSwitchStatement.class).firstOrThrow();
        assertFalse(switchStmt.hasDefaultCase());
        assertFalse(switchStmt.isExhaustiveEnumSwitch());
    }

    @Test
    void testEnumWithDefault() {
        @NonNull ASTSwitchStatement switchStmt = java.parse(ENUM_SWITCH_WITH_DEFAULT).descendants(ASTSwitchStatement.class).firstOrThrow();
        assertTrue(switchStmt.hasDefaultCase());
        assertFalse(switchStmt.isExhaustiveEnumSwitch());
    }

    private static final String SWITCH_WITH_DEFAULT =
        "public class Foo {\n"
            + " void bar() {\n"
            + "  switch (x) {\n"
            + "   default: y = 2;\n"
            + "   case 4: break;\n"
            + "  }\n"
            + " }\n"
            + "}";

    private static final String EXHAUSTIVE_ENUM =
        "public class Foo {\n"
            + " void bar() {\n"
            + "  enum LocalEnum { A, B, C } "
            + "  var v = LocalEnum.A; "
            + "  switch (v) {\n"
            + "   case A: break;\n"
            + "   case B: break;\n"
            + "   case C: break;\n"
            + "  }\n"
            + " }\n"
            + "}";

    private static final String NOT_EXHAUSTIVE_ENUM =
        "public class Foo {\n"
            + " void bar() {\n"
            + "  enum LocalEnum { A, B, C } "
            + "  var v = LocalEnum.A; "
            + "  switch (v) {\n"
            + "   case A: break;\n"
            + " //  case B: break;\n"
            + "   case C: break;\n"
            + "  }\n"
            + " }\n"
            + "}";

    private static final String ENUM_SWITCH_WITH_DEFAULT =
        "public class Foo {\n"
            + " void bar() {\n"
            + "  enum LocalEnum { A, B, C } "
            + "  var v = LocalEnum.A; "
            + "  switch (v) {\n"
            + "   case A: break;\n"
            + "   case C: break;\n"
            + "   default: break;\n"
            + "  }\n"
            + " }\n"
            + "}";
}

File: pmd-java/src/test/kotlin/net/sourceforge/pmd/lang/java/ast/ASTSwitchLabelTest.kt
/*
 * BSD-style license; for more info see http://pmd.sourceforge.net/license.html
 */

package net.sourceforge.pmd.lang.java.ast

import io.kotest.matchers.shouldBe
import net.sourceforge.pmd.lang.test.ast.shouldBe

/**
 * @author Clément Fournier
 */
class ASTSwitchLabelTest : ProcessorTestSpec({

    parserTestContainer("Different label forms") {
        inContext(ExpressionParsingCtx) {
            """
              switch (x) {
               default: y = 2;
               case null, default: y = 3;
               case null: y = 4;
               case 4: break;
               case 4, 6, a?b:c : break;
              }
            """.trimIndent() should parseAs {
                switchExpr {
                    it::hasDefaultCase shouldBe true
                    it::isExhaustiveEnumSwitch shouldBe false

                    variableAccess("x")
                    switchFallthrough {
                        switchDefaultLabel()
                        exprStatement()
                    }
                    switchFallthrough {
                        switchNullLabel(isDefault = true)
                        exprStatement()
                    }
                    switchFallthrough {
                        switchNullLabel(isDefault = false)
                        exprStatement()
                    }
                    switchFallthrough {
                        switchLabel {
                            int(4)
                        }
                        breakStatement()
                    }
                    switchFallthrough {
                        switchLabel {
                            it::isPatternLabel shouldBe false
                            it::getGuard shouldBe null
                            int(4)
                            int(6)
                            ternaryExpr()
                        }
                        breakStatement()
                    }
                }
            }
        }
    }

    parserTestContainer("Guards and patterns") {
        inContext(ExpressionParsingCtx) {
            """
              switch (x) {
               case Integer i, Label _: 
               case Label _: 
               case Record(int _): 
               case Integer i when i != 0: 
              }
            """.trimIndent() should parseAs {
                switchExpr {
                    it::hasDefaultCase shouldBe false
                    it::isExhaustiveEnumSwitch shouldBe false

                    variableAccess("x")
                    switchFallthrough {
                        switchLabel {
                            it.exprList.toList() shouldBe emptyList()
                            it::isPatternLabel shouldBe true
                            typePattern {
                                modifiers()
                                classType("Integer")
                                variableId("i")
                            }
                            typePattern {
                                modifiers()
                                classType("Label")
                                variableId("_")
                            }
                        }
                    }
                    switchFallthrough {
                        switchLabel {
                            it.exprList.toList() shouldBe emptyList()
                            it::isPatternLabel shouldBe true
                            typePattern {
                                modifiers()
                                classType("Label")
                                variableId("_")
                            }
                        }
                    }
                    switchFallthrough {
                        switchLabel {
                            it.exprList.toList() shouldBe emptyList()
                            it::isPatternLabel shouldBe true

                            recordPattern("Record")
                        }
                    }
                    switchFallthrough {
                        switchLabel {
                            it.exprList.toList() shouldBe emptyList()
                            it::isPatternLabel shouldBe true
                            typePattern {
                                modifiers()
                                classType("Integer")
                                variableId("i")
                            }
                            it::getGuard shouldBe child<ASTGuard> {
                                it::getGuard shouldBe infixExpr(BinaryOp.NE)
                            }
                        }
                    }

                }
            }
        }
    }
    class SwitchLabelParsingContext(arrow: Boolean) : NodeParsingCtx<ASTSwitchLabel>("switch label") {
        private val sep = if (arrow) " ->" else " :"

        override fun getTemplate(construct: String, ctx: ParserTestCtx): String =
            ExpressionParsingCtx.getTemplate(
                """
                    switch (x) {
                        $construct $sep 
                    }
                """.trimIndent(),
                ctx
            )

        override fun retrieveNode(acu: ASTCompilationUnit): ASTSwitchLabel =
            ExpressionParsingCtx.retrieveNode(acu)
                .let { it as ASTSwitchExpression }
                .branches[0]!!.label!!
    }

    parserTestContainer("Invalid syntax") {
        inContext(SwitchLabelParsingContext(arrow = false)) {
            "case Integer i, 4" shouldNot parse()
            "case 4, Integer i" shouldNot parse()
            "case null, null" shouldNot parse()
            "case null, default, null" shouldNot parse()
            "case default, null" shouldNot parse()
            "case 4 when 4 == 2" shouldNot parse()
            "case Integer i when 4 == 2, Boolean b when b" shouldNot parse()
            "case Integer i when 4 == 2, Boolean b" shouldNot parse()

            "case Integer i when 4 == 2" should parse()
            "case Integer i, Boolean b when b" should parse()
        }
    }

})
